Make the javalib independent of the Scala library.

sjrd opened the pull request:
Basic test suite passes in 2.12 and 2.13. Let's see how far this goes. 😆

sjrd commented:
Looks pretty good so far:
* All tests pass on 2.12 and 2.13
* Code size has improved
* 2.11 crashes the compiler during compilation, but that should be fixable

🙂

gzm0 commented:
> 2.11 crashes the compiler during compilation, but that should be fixable

Maybe it's time to ditch 2.11? :innocent:

sjrd commented:
> > 2.11 crashes the compiler during compilation, but that should be fixable
> 
> Maybe it's time to ditch 2.11? innocent

I suspect that, even when 2.11 compiles again, the IR cleaner will hit on 2.11-compiled code anyway, because 2.11 doesn't optimize `s"..."` interpolators as simple string concatenation, so it contains references to `StringContext`.

That said, ultimately, it won't matter whether 2.11 can compile our javalib or not, because we will only have to compile it once with one version (like `linkerPrivateLibrary`, which we only compile with 2.12). But there is a transient state where we need to make 2.11 "work" somehow, before we change the build to only compile the javalib with 2.12.

Unless we do the version merge at the same time as we enable the IR checker, but that would be a bit of a Frankenstein commit.

sjrd commented:
It's green! 😲 Now to redecompose all that into meaningful commits... 😄

sjrd commented:
Based on #4697, but otherwise ready to review :)

gzm0 commented on javalib/src/main/scala/java/util/regex/PatternCompiler.scala:
```diff
@@ -80,15 +82,15 @@ private[regex] object PatternCompiler {
   /** Cache for `Support.supportsUnicode`. */
   private val _supportsUnicode =
-    (esVersion >= ESVersion.ES2015) || featureTest("u")
+    (linkingInfo.esVersion >= ESVersion.ES2015) || featureTest("u")
```
Any reason not to import this?

sjrd replied:
Yes: `linkingInfo` is a `def`, so we cannot import members from it. I first tried an alias def, but I couldn't put it in `PatternCompiler` itself without destroying the constant folding (because `object PatternCompiler` is not pure), and it didn't feel right anywhere else.

sjrd commented on project/JavalibIRCleaner.scala:
```diff
@@ -552,14 +572,26 @@ object JavalibIRCleaner {
   private val jsAnyFromStringMethodName =
     MethodName("fromString", List(ClassRef(BoxedStringClass)), ClassRef(JSAny))
+  private val jsArrayOpsToArrayMethodName =
+    MethodName("jsArrayOps", List(ClassRef(JSArray)), ClassRef(JSArray))
```
It's this one:
https://github.com/scala-js/scala-js/blob/33ed7588ad80eaff5802aaf3b6f620b408a200cd/library/src/main/scala-old-collections/scala/scalajs/js/Any.scala#L92-L93
Because `js.ArrayOps` is an `AnyVal(js.Array)`, scalac erases this method to `(js.Array): js.Array)`. It's not a bridge.

sjrd commented on project/JavalibIRCleaner.scala:
```diff
@@ -664,4 +671,35 @@ object JavalibIRCleaner {
+   *  anymore after our erasure. We have to remove them, otherwise they become
+   *  duplicate definitions.
+   */
+  private val SupersededBridges: Set[(ClassName, MethodName)] = {
```
I had considered it and discarded the idea when I realized I would need at least as much code to make the generic version than to spell them out. But I can try again if you prefer.

gzm0 replied:
> But I can try again if you prefer.

Yes, please. Even if it is a bit more code, my hope is that it'll be easier to reason that the code is correct. In the current state, only the code here and the implementation of the relevant methods combined allow to determine if the implementation is correct.

sjrd replied:
OK, I have replaced the hard-coded list with generic logic, now.

gzm0 commented on javalib/src/main/scala/java/util/internal/MatchError.scala:
```diff
@@ -0,0 +1,31 @@
+package java.util.internal
+
+/** A replacement for `scala.MatchError`. */
+final class MatchError(obj: Any) extends RuntimeException {
```
Yes, I think that would be a good idea. Or even more minimalistic: It seems `AssertionError` has an `Object` constructor (https://github.com/scala-js/scala-js/blob/8b26ad30f5a820548da315ad57f833c53184b712/javalanglib/src/main/scala/java/lang/Throwables.scala#L225). So maybe we could just replace `MatchError` with `AssertionError`?

sjrd replied:
Directly replacing by `AssertionError` works :)

gzm0 commented on javalib/src/main/scala/scala/scalajs/js/typedarray/TypedArrayBufferBridge.scala:
```diff
@@ -36,45 +41,45 @@ package scala.scalajs.js.typedarray
 import java.nio._
 
 private[typedarray] object TypedArrayBufferBridge {
```
I've been thinking about this and from what you say, IIUC, we have no choice but to keep the interface and the implementation separate. However, if that is the case, why not use a `.java` file for the interface (rather than a non-machine readable list)?

That'll at least allow usage sites to benefit of type checking. (and IIUC the use case for separating the javalib is having more than one usage site).

sjrd commented:
Updated. I believe I have addressed all the comments that directly target this PR.

Do you want to resolve the discussion about the interface of the javalib for this PR, or should we continue discussing it separately?

gzm0 commented on project/JavalibIRCleaner.scala:
```diff
@@ -657,7 +657,11 @@ object JavalibIRCleaner {
 
-    val allPairs = functionTypePairs ++ refPairs ++ tuplePairs
+    val otherPairs = List(
+      ClassName("scala.MatchError") -> ClassName("java.lang.AssertionError"),
```
Consider putting the commit comment as code comment here.

> `AssertionError` conveniently features a constructor taking an
`Object`. Since any `MatchError` in the javalib would be a bug, it
is fine to rewrite them to `AssertionError`s.

sjrd replied:
Good point. Done.

gzm0 reviewed (approved):
Just a nit RE commenting of AssertionError.

> Do you want to resolve the discussion about the interface of the javalib for this PR, or should we continue discussing it separately?

We can discuss separately. We shouldn't block this PR on the outcome of the discussion IMO (it's clear to me that there are viable options, so we can decide on the specific one when we need it).
